File: Platonov/365.cpp
class Solution {
public:
    bool canMeasureWater(int jug1Capacity, int jug2Capacity, int targetCapacity) {
        if (jug1Capacity > jug2Capacity) {
            swap(jug1Capacity, jug2Capacity);
        }
        queue<pair<int, int> > q;
        map<pair<int, int>, bool > used;
        used[make_pair(0, 0)] = true;
        q.push(make_pair(0, 0));
        pair<int, int> coords;
        int difference, x, y;
        pair<int, int> newCoords;
        while (!q.empty()) {
            coords = q.front();
            x = coords.first;
            y = coords.second;
            if (x == targetCapacity || y == targetCapacity || x + y == targetCapacity) return true;
            q.pop();
            
            if (x + y - jug1Capacity > 0)
                difference = x + y - jug1Capacity;
            else difference = 0;

            newCoords = make_pair(x + y - difference, difference);
            
            if (used.find(newCoords) == used.end()) {
                used[newCoords] = true;
                q.push(newCoords);
            }

            if (x + y - jug2Capacity > 0)
                difference = x + y - jug2Capacity;
            else difference = 0;

            newCoords = make_pair(difference, y + x - difference);
            
            if (used.find(newCoords) == used.end()) {
                used[newCoords] = true;
                q.push(newCoords);
            }

            newCoords = make_pair(jug1Capacity, y);

            if (used.find(newCoords) == used.end()) {
                used[newCoords] = true;
                q.push(newCoords);
            }

            newCoords = make_pair(x, jug2Capacity);

            if (used.find(newCoords) == used.end()) {
                used[newCoords] = true;
                q.push(newCoords);
            }

            newCoords = make_pair(0, y);

            if (used.find(newCoords) == used.end()) {
                used[newCoords] = true;
                q.push(newCoords);
            }

            newCoords = make_pair(x, 0);

            if (used.find(newCoords) == used.end()) {
                used[newCoords] = true;
                q.push(newCoords);
            }
        }
        return false;
    }
};

File: Nenakhov/365.py
class Solution(object): 
    def canMeasureWater(self, jug1Capacity, jug2Capacity, targetCapacity):
        """
        :type jug1Capacity: int
        :type jug2Capacity: int
        :type targetCapacity: int
        :rtype: bool
        """
        my_stack = collections.deque()
        my_stack.append((0,0))
        memory = set((0,0))
        #пройдемся DFS-ом
        while my_stack:
            arr = my_stack.pop()
            a = arr[0]
            b = arr[1]
            if a+b == targetCapacity:
                return True
            #смотрим все возможные дальнейшие операции, которые доступны из условия
            for i in [(jug1Capacity,b), (a,jug2Capacity), (0,b), (a,0), (min(a+b,jug1Capacity), max(b-(jug1Capacity-a),0)), (max(a-(jug2Capacity-b),0),min(a+b,jug2Capacity))]:
                if i not in memory:
                    memory.add(i)
                    my_stack.append(i)
        return False 
